torch/csrc/autograd/custom_function.cpp

```diff
@@ -427,15 +443,21 @@ optional_variable_list _wrap_outputs(
     const std::unordered_set<at::TensorImpl*>& dirty_inputs,
     const at::ArrayRef<c10::optional<Variable>> raw_outputs,
     const std::shared_ptr<Node>& cdata,
-    _jvp_fn_t jvp_user_function) {
+    _jvp_fn_t jvp_user_function,
+    const std::unordered_set<at::TensorImpl*>& to_save_if_setup_context) {
   std::unordered_map<at::TensorImpl*, size_t> inputs_mapping;
   inputs_mapping.reserve(input_vars.size());
   for (const auto i : c10::irange(input_vars.size())) {
     inputs_mapping.emplace(input_vars[i].unsafeGetTensorImpl(), i);
   }
 
   auto outputs = _process_backward_mode_ad(
-      inputs_mapping, non_differentiable, dirty_inputs, raw_outputs, cdata);
+      inputs_mapping,
+      non_differentiable,
+      dirty_inputs,
+      raw_outputs,
+      cdata,
+      to_save_if_setup_context);
```

Comment:
As another sanity check: It's not a problem for forward-mode AD if this happens (i.e. there is no ambiguity if we save_for_forward an input that has been returned as an output), right? Should we add a test for that?

Reply:
I'd need to check the functorch behavior, but either way I'd happy with also disallowing save_for_forward on inputs saved as outputs when used with setup-context as well.

Reply:
Test added.

Reply:
Ugh I think we may need to disallow save_for_forward as well
```python
class A(torch.autograd.Function):
    @staticmethod
    def forward(x):
        return x

    @staticmethod
    def setup_context(ctx, inputs, output):
        ctx.save_for_forward(output, inputs[0])

    @staticmethod
    def jvp(ctx, x_t):
        out, inp = ctx.saved_tensors
        # It does matter whether you use out, inp here!
        return x_t * out

x = torch.tensor(2., device=device)
x_t = torch.tensor(2., device=device)

# If we do just jvp, everything is OK
jvp(A.apply, (x,), (x_t,))

def jvp_fn(fn):
    def inner(x, x_t):
        return jvp(fn, (x,), (x_t,))[1]
    return inner

# There is a difference between using inp/out in the grad(jvp case
# If we use inp, the grad( does not need to backprop through any
# ABackward nodes
# If we use out, we would error out here because the out has
# ABackward as its grad_fn and it expects the vjp to be
# implemented 
grad(jvp_fn(A.apply))(x, x_t)
```

Reply:
It is now updated to do so.

Reply:
Just to check, grad(vmap( is fine, right? I'd expect that case to be fine because vmap would preserve the identity of the input

Reply:
If you mean vmap(grad, then yes. (but it doesn't matter that vmap preserves identity because the autograd Function is still applied to the inner-most tensor and that aliases always, so the fix [below ](https://github.com/pytorch/pytorch/pull/97212#discussion_r1148046522) also fixes that.

Reply:
I was talking about grad(vmap:
```
import torch
from torch.func import grad

class Cube(torch.autograd.Function):
    generate_vmap_rule = True

    @staticmethod
    def forward(x):
        return x ** 3, x

    @staticmethod
    def setup_context(ctx, inputs, outputs):
        cube, x = outputs
        ctx.save_for_backward(x)

    @staticmethod
    def backward(ctx, grad_output, grad_x):
        # NB: grad_x intentionally not used in computation
        x, = ctx.saved_tensors
        result = grad_output * 3 * x ** 2
        return result


def f(x):
    cube, inp = Cube.apply(x)
    return cube

x = torch.tensor([1., 2., 3.], requires_grad=True)

print("============= functorch")
grad_x = grad(lambda x: torch.vmap(f)(x).sum())(x)
print(grad_x)
```
(which does error out correctly with this PR!)

Reply:
Oh ok, I see what you mean now. Nice!